mypy/newsemanal/semanal_typeargs.py

```diff
@@ -0,0 +1,79 @@
+"""Verify that type arguments like 'int' C[int] are valid.
+
+This must happen after semantic analysis since there can be placeholder
+types until the end of semantic analysis, and these break various type
+operations, including subtype checks.
+"""
+
+from typing import List
+
+from mypy.nodes import TypeInfo, Context, MypyFile, FuncItem, ClassDef, Block, OverloadedFuncDef
+from mypy.types import Type, Instance, TypeVarType, AnyType
+from mypy.mixedtraverser import MixedTraverserVisitor
+from mypy.subtypes import is_subtype
+from mypy.sametypes import is_same_type
+from mypy.errors import Errors
+from mypy.scope import Scope
+from mypy import message_registry
+
+
+class TypeArgumentAnalyzer(MixedTraverserVisitor):
+    def __init__(self, errors: Errors) -> None:
+        self.errors = errors
+        self.scope = Scope()
+
+    def visit_mypy_file(self, o: MypyFile) -> None:
+        self.errors.set_file(o.path, o.fullname(), scope=self.scope)
+        self.scope.enter_file(o.fullname())
+        super().visit_mypy_file(o)
+        self.scope.leave()
+
+    def visit_func(self, defn: FuncItem) -> None:
+        with self.scope.function_scope(defn):
+            super().visit_func(defn)
+
+    def visit_class_def(self, defn: ClassDef) -> None:
+        with self.scope.class_scope(defn.info):
+            super().visit_class_def(defn)
+
+    def visit_block(self, o: Block) -> None:
+        if not o.is_unreachable:
+            super().visit_block(o)
+
+    def visit_instance(self, t: Instance) -> None:
+        info = t.type
+        for (i, arg), tvar in zip(enumerate(t.args), info.defn.type_vars):
+            if tvar.values:
+                if isinstance(arg, TypeVarType):
+                    arg_values = arg.values
+                    if not arg_values:
+                        self.fail('Type variable "{}" not valid as type '
+                                  'argument value for "{}"'.format(
+                                      arg.name, info.name()), t)
+                        continue
+                else:
+                    arg_values = [arg]
+                self.check_type_var_values(info, arg_values, tvar.name, tvar.values, i + 1, t)
+            if not is_subtype(arg, tvar.upper_bound):
+                self.fail('Type argument "{}" of "{}" must be '
+                          'a subtype of "{}"'.format(
+                              arg, info.name(), tvar.upper_bound), t)
+        super().visit_instance(t)
```

Comment:
The old third pass not just checked the argument types, but also argument counts. For example it replaced `List[int, str]` with `List[Any]` and emitted an error. Why don't we also do this here?

Reply:
We do this earlier in the type analyzer, since the number of type variables is known earlier. I'll add a comment here.

Reply:
> We do this earlier in the type analyzer, since the number of type variables is known earlier.

Well, it looks like we don't do this for type aliases, see
https://github.com/python/mypy/blob/master/mypy/newsemanal/typeanal.py#L350. I will open an issue for this.